### PR TITLE
Add BTI support for AArch64 RSEQ implementation

### PR DESCRIPTION
**Pull request by avieira-arm**
Hi,

This patch adds support for BTI-enabled execution in the AArch64 RSEQ implementation.

Tested with:
CC=clang bazel test -c opt --copt "-mcpu=neoverse-n1" --linkopt "-fuse-ld=lld"  --sandbox_debug --override_repository="tcmalloc=src/tcmalloc" -k //tcmalloc/...
CC=clang bazel test -c dbg --copt "-mcpu=neoverse-n1" --linkopt "-fuse-ld=lld"  --sandbox_debug --override_repository="tcmalloc=src/tcmalloc" -k //tcmalloc/...

Note that the hardware I tested it on does not support BTI, so the 'BTI C' instruction behaves as a nop in the following.
I used '--copt "-mbranch-protection=bti"' to compile all of tcmalloc with BTI enabled. However, because the bootcode/library code I'm using is not BTI enabled I used '--linkopt "-Wl,-z,force-bti" to force the inclusion of the .note.gnu.property section in the resulting binary.

Doing this will cause the linker to warn what files it had to forcefully add the note to, these were all non-tcmalloc files. Tests pass and tcmalloc_benchmark runs to completion. This leads me to believe that with this patch TCMalloc is BTI-ready. 

Is this OK?

Kind Regards,
Andre

### CODE REVIEW

**Comment by zbrid** [comment]
@ckennelly @resistor Hi Chris and Owen,

Do we have an ARM machine with BTI that one of you could try this patch on for a sanity check that it works? Looks like Andre didn't have access to one.

**Comment by zbrid** [comment]
@ckennelly @avieira-arm Is the AArch64 RSEQ implementation the only part of TCMalloc that needs to be mititgated or do we expect more uses of BTI to be added elsewhere?

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (zbrid) [inline tcmalloc/internal/percpu_rseq_aarch64.S:172]: How is PAC support related to BTI here? I'm not sure how this TODO relates to the rest of the patch, so asking to learn here.

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (zbrid) [inline tcmalloc/internal/percpu_rseq_aarch64.S:522]: I haven't seen the use of gnu properties before, so I had a question about this: why did you choose these values for the type and value?

**Review by zbrid — CHANGES_REQUESTED** [review]
Thanks for putting up this PR, Andre!

Is there a missing BTI_C on line 355 in the original file? If not, why is that function different and would it be useful to put a comment describing why BTI_C isn't needed there?

I also had a few questions inline. Thanks!

**Comment by ckennelly** [comment]
> @ckennelly @avieira-arm Is the AArch64 RSEQ implementation the only part of TCMalloc that needs to be mititgated or do we expect more uses of BTI to be added elsewhere?

All of the other assembly is generated by the compiler, so it can take care of inserting BTI as required.

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (avieira-arm) [reply tcmalloc/internal/percpu_rseq_aarch64.S:172]: PAC(Pointer Authentication) is not directly related to BTI. They are however often used together as PAC is meant to protect against ROP attacks and BTI against JOP attacks. Having said that, this was intended as a helpful reminder that in this particular case, because we potentially spill LR we might want to protected it using PAC.

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (avieira-arm) [reply tcmalloc/internal/percpu_rseq_aarch64.S:522]: I hadn't seen it before either, but IIUC this is added such that the object file resulting from the assembly of this file contains a .note.gnu.property NOTE section, and these values indicate the code is BTI enabled. The linker, by default, checks whether all included object files contain such a note. If that is the case then the resulting binary will also be marked as BTI-enabled, which I believe means that the loader will then set the appropriate tag to indicate it is a 'guarded memory region'.

This is why I used '-z force-bti' as then the linker will insert the note section in the final binary regardless and warn us about what object files were missing it. In my case those were all boot-up code that had not been built with BTI enabled. For C/C++ code the compiler will create the note for you if you tell it to enable BTI.

As to answer the original question as to why these values, I looked at other uses of this note section in assembly to get the approrpriate values, for instance here:
https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=libgcc/config/aarch64/lse.S;h=7496b8b6337fb2c924ba338046df396522aa0a85;hb=refs/heads/master#l281

**Comment by avieira-arm** [comment]
Yeah there is a BTI_C missing there. I'll add it. Thanks!

Regarding testing, as far as I am aware there is no BTI-enabled hardware yet, but I have been told I may be able to test this with qemu. Though this doesn't seem to be simple. Maybe its worth postponing proper testing of this until hardware is available. I will see how far I get with qemu.

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (zbrid) [reply tcmalloc/internal/percpu_rseq_aarch64.S:522]: Thanks for that very helpful explanation!

**Review by zbrid — APPROVED** [review]
@avieira-arm Thanks for all the answers! Also thanks for looking into trying with qemu.

For the testing portion, I'll leave the decision on whether to postpone upstreaming of this until hardware is available to @ckennelly since he owns TCMalloc.

From the security team side, I think this patch looks good once that last BTI_C is added on 355. I'll LGTM with the understanding that'll be added before merging.

File: tcmalloc/internal/percpu_rseq_aarch64.S
Comment (ckennelly) [reply tcmalloc/internal/percpu_rseq_aarch64.S:523]: I was puzzled by the other note, but it seems like both are needed ([based on this similar LLVM commit](https://reviews.llvm.org/rGd53e61863d48a07ce285d5b0a36abc67583023bd))

**Review by ckennelly — COMMENTED** [review]
Let's rebase, since this predates https://github.com/google/tcmalloc/pull/65

**Comment by avieira-arm** [comment]
> Let's rebase, since this predates #65

So #65 doesn't need any changes for BTI since the compiler should generate the BTI instructions if it doesn't choose to inline the new functions. However, I did rebase the assembly file to match the new labels that now have 'Internal' in their name.

**Review by ckennelly — APPROVED** [review]

**Pull request merged**
